### PR TITLE
Dashboard: Migration - EditVariable Settings: Implement DataSource Variable 

### PR DESCRIPTION
**Pull request by axelavargas**
https://github.com/grafana/grafana/assets/239999/29bdbd61-9200-4031-ad46-1c5ef25b0ab4

- [x] Extract common logic from DataSourceVariableEditor into a new component
- [x] Use new logic in core grafana
- [x]  Use new logic in scenes grafana
- [x] Add unit test
- [x] (Fixed in another PR) - Fix issue when regex triggers an error, this happens when there are no data source options to show in Preview Values (Figure out how we can handle errors in this form)

Part of [#78690](https://github.com/grafana/grafana/issues/78690)

### CODE REVIEW

File: public/app/features/variables/datasource/DataSourceVariableEditor.tsx [hunk 1]
```diff
@@ -93,49 +87,29 @@ export class DataSourceVariableEditorUnConnected extends PureComponent<Props> {
...
+        optionTypes={typeOptions}
+        onChange={this.onDataSourceTypeChanged}
+        onRegExBlur={this.onRegExBlur}
+        onMultiChange={this.onMultiChanged}
+        onIncludeAllChange={(event) =>
+          this.onSelectionOptionsChange({ propName: 'includeAll', propValue: event.currentTarget.checked })
+        }
+        onAllValueChange={(event) =>
+          this.onSelectionOptionsChange({ propName: 'allValue', propValue: event.currentTarget.value })
+        }
```
Comment (ivanortegaalba) [inline public/app/features/variables/datasource/DataSourceVariableEditor.tsx:111]: Small nit: Probably we can move these two handlers to their own instance methods like `onMultiChanged`. Just for consistency. WDYT?

**Review by ivanortegaalba — COMMENTED** [review]
This looks great Alexa! Let's adds some tests, and we can get this merge 💪

File: public/app/features/dashboard-scene/settings/variables/utils.test.ts [hunk 2]
```diff
@@ -27,12 +30,44 @@ import {
...
+  uid: SHARED_DASHBOARD_QUERY,
+  getRef: () => {
+    return { type: SHARED_DASHBOARD_QUERY, uid: SHARED_DASHBOARD_QUERY };
+  },
+} as DataSourceApi<DataQuery, DataSourceJsonData, {}>;
+
+jest.mock('@grafana/runtime', () => ({
+  ...jest.requireActual('@grafana/runtime'),
+  getDataSourceSrv: () => ({
+    get: async () => dsMock,
```
Comment (axelavargas) [inline public/app/features/dashboard-scene/settings/variables/utils.test.ts:55]: @dprokop this did the trick to fix the unit test 😄

Comment (axelavargas) [reply public/app/features/dashboard-scene/settings/variables/utils.test.ts:55]: @ivanortegaalba, This is relevant for your Query Editor Migration PR, the tests are failing for the same reason 😄 https://drone.grafana.net/grafana/grafana/155958/1/6

File: public/app/features/variables/datasource/DataSourceVariableEditor.test.tsx [hunk 3]
```diff
@@ -45,10 +46,21 @@ describe('DataSourceVariableEditor', () => {
...
+    await user.click(field);
+    await user.type(field, '/prod/');
+    expect(field).toHaveValue('/prod/');
+    await user.tab();
+    expect(props.onPropChange).toHaveBeenCalledWith({ propName: 'regex', propValue: '/prod/', updateOptions: true });
   });
 });
+
+// based on styleguide recomendation
+function setup(jsx: JSX.Element) {
```
Comment (ivanortegaalba) [inline public/app/features/variables/datasource/DataSourceVariableEditor.test.tsx:61]: 🤔 TIL

**Review by ivanortegaalba — APPROVED** [review]
Amazing work Alexa! 👏 

✅ Components reused
✅ Old version working and test updated
✅ New components updating variable state

All covered! 💯

File: public/app/features/dashboard-scene/settings/variables/utils.ts [hunk 4]
```diff
@@ -140,3 +143,18 @@ export function getDefinition(model: SceneVariable): string {
...
+  const datasources = getDataSourceSrv().getList({ metrics: true, variables: true });
+
+  const optionTypes = chain(datasources)
+    .uniqBy('meta.id')
+    .map((ds: DataSourceInstanceSettings) => {
+      return { label: ds.meta.name, value: ds.meta.id };
+    })
+    .value();
+
+  optionTypes.unshift({ label: '', value: '' });
```
[the remark this answers is not in the document]
Comment (axelavargas) [reply public/app/features/dashboard-scene/settings/variables/utils.ts:157]: @torkelo The old arch included a blank option [code](https://github.com/grafana/grafana/blob/main/public/app/features/variables/datasource/actions.ts#L59) , and I chose to maintain this behavior when migrating.

[the remark this answers is not in the document]
Comment (torkelo) [reply public/app/features/dashboard-scene/settings/variables/utils.ts:157]: @axelavargas strange, it creates a buggy behavior of no placeholder text showing and a ugly empty line in the select.

Fixed in https://github.com/grafana/grafana/pull/104153

**Pull request merged**